Building visp_java module

AKS1996 opened the pull request:
+ Equivalent to the `opencv_java` module - providing basic conversion functions and the JNI support too.
+ Made a few edits. Will soon run the tests and provide results. 
+ Yet to be edited: android_test folder, CMakeLists.txt, generators/src folder, some files in pure_test folder

AKS1996 commented:
Hi @fspindle  @s-trinh
 
Need to define some java files as [of OpenCV](https://github.com/opencv/opencv/tree/master/modules/core/misc/java) for the java module. But their directory structure is different from that of ViSP. Should separate the files in `core/src` and `core/test`?

fspindle commented:
If it helps, we could introduce `core/misc/java/src` and `core/misc/java/test` as OpenCV

AKS1996 commented:
Do `vpMatrix`, `vpRect`, `vpPoint` support any data type other than `double`? Coz then there will be fewer functions for usage in java.

AKS1996 commented:
@fspindle 
Should I keep class name qualifiers as CamelCase -  `VpMatrix`/`VpException` instead of `vpMatrix`/`vpException` in java?

fspindle commented:
For me `vpMatrix`, `vpRect`, `vpPoint` should support only double as it is by now.

Concerning the class name, could you give us more details. I don't see why this is needed and didn't find similar things in OpenCV.

AKS1996 commented:
@fspindle @s-trinh 
Recently I made some edits in `visp_java` module. The build ran successfully on my device but Travis is failing([Refer here](https://travis-ci.org/lagadic/visp/jobs/378006921#L2789)). Am I missing some code or header file? Does Travis CI has necessary JNI library?

AKS1996 commented:
hi @fspindle @s-trinh 
Do you have Travis configured to provide NDK support? I can't find sign of either JNI or NDK in this PR or in #307. You can refer [this thread](https://stackoverflow.com/questions/38090358/travis-how-to-know-android-sdk-ndk-path) that the SDK that Travis provides by default is old/buggy. To get the support, [refer this](https://stackoverflow.com/questions/49666268/how-to-setup-travis-ci-for-an-android-project-which-uses-both-ndk-and-sdk-to-bui).

fspindle commented:
Hi @AKS1996 

We were in vacation last week and busy yesterday. I will check for travis today

fspindle commented:
@AKS1996 

- I did some [changes](https://github.com/lagadic/visp/pull/323/commits/d466df955f1deaea00397e51b3227f9dbca8b032) but to be able to enable and build the `visp_java`module, I think that in the main `CMakeList.txt` there should be something to detect `ant` as for OpenCV (see [here](https://github.com/opencv/opencv/blob/master/CMakeLists.txt#L300)):
````
VP_OPTION(BUILD_JAVA               "Enable Java support"                         (ANDROID OR NOT CMAKE_CROSSCOMPILING)  IF (ANDROID OR (NOT APPLE_FRAMEWORK AND NOT WINRT)) )
````
and then like [here](https://github.com/opencv/opencv/blob/master/CMakeLists.txt#L638)

````
# --- Java Support ---
if(BUILD_JAVA)
  if(ANDROID)
    include(cmake/android/VISPDetectAndroidSDK.cmake)
  else()
    include(cmake/VISPDetectApacheAnt.cmake)
    find_package(JNI)
  endif()
endif()
````

- It could be also interesting, before having a full Android support, to validate the java module and to make tutorials showing how to build ViSP for Java (see [here](http://opencv-java-tutorials.readthedocs.io/en/latest/01-installing-opencv-for-java.html)) and how to create the first Java app with ViSP (see [here](http://opencv-java-tutorials.readthedocs.io/en/latest/02-first-java-application-with-opencv.html))

fspindle commented:
In Travis and Appveyor the build (without NDK support right now) is working but the tests are failing because all the binaries are located in `bin/android` due to this [line](https://github.com/AKS1996/visp/blob/java_module/CMakeLists.txt#L705). Removing the line should make ci working.

AKS1996 commented:
Hi @fspindle

Nice point regarding the first Java app with ViSP. This way, as you already mentioned, we can check validity of `visp_java` module first, then later switch to an `android_build`.

I'll push the changes you recommended and update you with a demo java app soon.

AKS1996 commented:
Hi @fspindle 

Sorry for responding so late. I was messed up in setting eclipse for JDK development. Nevertheless, I ran [an example](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/JNI-Usage-Example) of using visp in java. Currently the code is standalone - I'm trying to figure out how to use `visp_java` module for it.

Akshay

AKS1996 commented:
Hi @fspindle 
Do have a look it this [this example](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/JNI-Usage-Example). I'm sure it's the same way how `opencv_java`([refer this](https://github.com/opencv/opencv/blob/master/modules/core/misc/java/src/java/core%2BMat.java) for more - the file structure is similar). I'm learning how to manage dependencies using `ant`(rather than a compile script).

I see [on docs](http://visp-doc.inria.fr/doxygen/visp-daily/classvpMatrix.html) that `vpMatrix` class contains ~100 member functions. So,
+ First, I'll try to add more Java functions to `org.visp.core.VpMatrix` class
+ Then I'll shift back to the `ant` build procedure

Thanks
Akshay

AKS1996 commented:
Hi @fspindle @s-trinh 

I see in [opencv's java module](https://github.com/opencv/opencv/tree/master/modules/core/misc/java/src/java) that they didn't include native support for all basic data structures. Only `opencv::Mat` uses the native `c++` code, rest all structures (like [`opencv::Point`](https://github.com/opencv/opencv/blob/master/modules/core/misc/java/src/java/core%2BPoint.java)) are coded in Java only.

Does it have to be the same with `ViSP`? I'm not sure why opencv has such structure - maybe they wanted to restrict the SDK or maybe other structures didn't need any native support. `ViSP` supports classes like `vpArrayDouble`, `vpImage`. Do I need to include native support for them?

Thanks
Akshay

s-trinh commented:
Hi,
First of all, sorry for my lack of knowledge on this topic (JNI, Java binding / wrapping, ...).
To try catch-up with the current state, I have some questions:

- how [`gen_java.py`](https://github.com/AKS1996/visp/blob/d466df955f1deaea00397e51b3227f9dbca8b032/modules/java/generator/gen_java.py) is working? I understand that this script will somehow allow creating automatically the binding / wrapping of the C++ classes and functions to have the same interface in Java. JNI is the low-level framework that allows calling C++ code from Java. Is it correct? Will it do that automatically for all the ViSP classes?

- "native" or custom C++ data-type is manually wrapped, e.g. `core+Point.java` I can assume the previous script can do it automatically for simple data structure like `cv::Point` but this is done manually to override specific Java methods (e.g. `hashCode()`, `equals()`, `toString()` to be able to use them with Java structures and algorithms). Also, in C++ we override operators like `operator+()` but we need to provide a function call in Java. Does my understanding correct?

- I can also assume OpenCV provides a manual JNI wrapping of `cv::Mat` since this is a complicated data-structure and to be able to correctly handle the memory management (`cv::Mat`) is a kind of ref-pointer under the hood. It makes sense to do the same for `vpMatrix` if it is needed. Probably `vpImage` will need a specific JNI wrapping?

- I would say provide a JNI wrapping + Java interface for data-structures that need it, otherwise simple Java interface similar to `core+Point.java` is fine for me.

Also, with OpenCV I can get a jar file and use Java OpenCV in Eclipse. Is it possible to do the same? Currently I am not able to build the java module, maybe because it needs the Android SDK?

AKS1996 commented:
I'm not sure about the second point. You're correct that Java doesn't provide operator for classes. So manual wrapping may provide functions for that.

AKS1996 commented:
Third point is actually what I was asking about. Native code should be used for `cv::Mat` due to memory management issues - so is true for `vpMatrix`. I wanted to ask which `ViSP` classes do you think need such support(already mentioned `vpImage`). Or which ViSP classes have memory intensive functions/operations? 
If we are not clear about this then we might need to add native support for such classes later - which will be difficult since debugging android apps that use C++ code is hard.

AKS1996 commented:
Regarding the opencv's jar file, I was trying to do the same for visp using Eclipse. I need some time to separate the java and android modules(I'll keep it in mind while future development too) for visp, for now they are mixed. Currently `visp_java` module is not building correctly. I'll update you about it soon

fspindle commented:
For me there are the `vpMatrix`, `vpArray2D` and `vpImage` classes that need such support.

s-trinh commented:
> Third point is actually what I was asking about. Native code should be used for cv::Mat due to memory management issues - so is true for vpMatrix. I wanted to ask which ViSP classes do you think need such support(already mentioned vpImage). Or which ViSP classes have memory intensive functions/operations?

Corresponding ViSP classes should be:

- classes that inherit from the template class `vpArray2D`, i.e. vpMatrix, vpColVector, ... (see also classes that are in this [folder](https://github.com/lagadic/visp/tree/master/modules/core/src/math/matrix) and maybe also [these](https://github.com/lagadic/visp/tree/master/modules/core/src/math/transformation) since they inherit from `vpArray2D`?)

- `vpSubColVector`, `vpSubRowVector`, `vpSubMatrix` can be omitted in my opinion (not really used inside ViSP and not useful from a user point of view in my opinion)

My suggestion should be to start to wrap classes that would allow using the `imgproc` module and `vpAprilTag` class (as stated in your [proposal](https://github.com/AKS1996/GSOC-Prep/blob/master/Proposal/Proposal.pdf)).

To achieve this, the following data-structure should need to be wrapped: `vpPoint`, `vpImagePoint`.

AKS1996 commented:
Hi @fspindle @s-trinh 

Do have a look [at this repo](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/JNI-Usage-Example)(in the `cpp` and `org/visp/core` folders respectively). The function definitions are almost similar to `opencv` ones. 
Let me know if I should change/remove generic data type support from `vpArray2D` or `vpimage`. `opencv` employs enumerations to restrict data types to a few selected ones - like `CV_23F`, etc. Should we do the same?

Thanks
Akshay

AKS1996 commented:
Should I add native support for `visp_gui` too? I just generated a call graph for april tag detection tutorial. Member function of the class `vpImage` and `vpDisplayOpenCV` were called the most. Else to show the demo in java(on linux), we need to think of some other graphics support library.

First I'lll try using an image, without a graphics console.

fspindle commented:
From my point of view, what should be supported
- In `visp_gui` only `vpDisplayOpencv`. Other existing classes will be not compatible. The goal would be to be able to display a `vpImage` and some basic drawings like crosses and lines. If time remains at the end of the GSoC we will see for other graphics libraries that may be used to avoid the conversion from `vpImage` to `cv::Mat` that is achieved internally
- all `visp_vision` to support `vpPose` and `vpHomography` classes
- in `visp_blob` only `vpDot2` class for the moment

Validation could be done with a demo/tutorial with 4 blobs organized as a square and pose estimation (see [tutorial](http://visp-doc.inria.fr/doxygen/visp-daily/tutorial-pose-estimation.html)). For this demo we will need to [calibrate](http://visp-doc.inria.fr/doxygen/visp-daily/tutorial-calibration-intrinsic.html) the camera (could be done on Ubuntu with images acquired by the smartphone). The distance in meter between 2 blobs should be also known. Then comes the critical part with the initial detection of the blobs that is requested to initialize the blob tracker. I think that we could use `vpDot2` [auto detection](http://visp-doc.inria.fr/doxygen/visp-daily/tutorial-tracking-blob.html#tracking_blob_auto) capabilities. Knowing the size of the blob, its gray level min/max we should be able to detect the 4 blobs when they are at a given distance from the camera. 

I will try to find time to validate this idea on Ubuntu and send you the code that could be used for the demo.

Concerning your previous post, I'm not so enthusiastic to introduce `VP_32F` and `VP_8U` for `vpImage` and `vpArray` classes. If native support of these classes is not possible without `VP_32F` and `VP_8U` let us know. As I remmember, `vpArray` is only used with `double`, while `vpImage` classe only with `unsigned char` and `double`.  In `vpImage` there is also the support of `short` type, but not really used. We can forget it for the moment.

AKS1996 commented:
Hi @fspindle @s-trinh 

Do have a look at this: it's a part of the output while generating the make files
```
--   ViSP modules:
--     To be built:                 core gui imgproc io java java_bindings_generator klt me sensor ar blob robot visual_features vs vision detection mbt tt tt_mi
--     Disabled:                    -
--     Disabled by dependency:      java
--     Unavailable:                 -
```
It's clear to me that I'm missing some dependency. But it's turning out hard to find which one. I guess some `add_custom_command` macro call will have a `DEPENDS` part that's missing. Do you know of any way(other than manually debugging the files) to find what's missing? 

Apart from that, almost half way adapting `gen_java.py` file for generating JNI and Java files for `visp`. I've added the `vpMatrix`, `vpImage`, `vpArray2D` manually for JNI support. Also, for now the `gui` module in java will support only member functions of `vpDisplayOpenCV.cpp` file.

I followed the [opencv java eclipse tutorial](http://opencv-java-tutorials.readthedocs.io/en/latest/01-installing-opencv-for-java.html). I built a small demo - refer [this repo](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/JNI-Usage-Example) for the sample `.jar` and `.so` file on linux. Tell me if the `jar` structure needs to be changed.

Akshay

fspindle commented:
Not sure, but it could be due to these lines in `java/CmakeLists.txt` where a 3rd party is missing

```
if(APPLE_FRAMEWORK OR WINRT OR NOT PYTHON_DEFAULT_AVAILABLE OR NOT ANT_EXECUTABLE
    OR NOT (JNI_FOUND OR (ANDROID AND ANDROID_NATIVE_API_LEVEL GREATER 7)))
  vp_module_disable(java)
endif()
```

s-trinh commented:
> Should I add native support for visp_gui too? I just generated a call graph for april tag detection tutorial. Member function of the class vpImage and vpDisplayOpenCV were called the most. Else to show the demo in java(on linux), we need to think of some other graphics support library.

Alternative could be to use the native Java GUI capabilities (e.g. Java Swing) and provide helpers, something similar to [`highgui+HighGui.java`](https://github.com/opencv/opencv/blob/master/modules/highgui/misc/java/src/java/highgui%2BHighGui.java) maybe? This should be just for display images. To draw lines, circles, ..., custom Java code could be wrapped into ready to use functions?

AKS1996 commented:
Thanks, that line in `java/CMakeLists.txt` was the issue. I'll add python support to visp later(for detecting the version, etc).

AKS1996 commented:
What's the purpose of `VISP_EXPORT` in classnames qualifiers? Like `class VISP_EXPORT vpXmlParserHomogeneousMatrix`

AKS1996 commented:
Found [this excellent](https://docs.opencv.org/3.0-beta/doc/py_tutorials/py_bindings/py_bindings_basics/py_bindings_basics.html) about generating opencv bindings in other languages(Python, Java, etc)

fspindle commented:
`VISP_EXPORT` is used mainly for windows to import/export classes in/from DLL libraries. OpenCV has a similar macro called `CV_EXPORTS`

AKS1996 commented:
Hi @fspindle @s-trinh 
I've added the patches for building `visp_java` module. As in `opencv`, you'll get a `visp320.jar` executable and corresponding `libvisp320.so` library. Do try to install and let me know of any issues/errors.

I've removed tests for now. I've adapted the `opencv`'s java module structure - only modules inside
`modules/java/misc` will have java support. Currently I've added a few basic classes for `core` module. 

Until the first evaluation, I'll try to add classes and member functions necessary for aprilTag demo i.e. `core` `imgproc`, `detection` and update you of any progress

Thanks
Akshay

AKS1996 commented:
Forgot to ask - are the CI tests for PR failing due to some build related issue?

AKS1996 commented:
In the last commits, I made some edits to `vp_add_module` macro.

To add the `java` wrapper for a module, edit this line in its CMakeLists.txt:
`vp_add_module(<mod-name> ...)`  to  `vp_add_module(<mod-name> ... WRAP java)`

AKS1996 commented:
Do we need generic support for `vpImage` and `vpArray2D` for java? Adding generics for java is easy but workarounds for generic class support in JNI will be confusing. Even opencv has separate [cv::Mat types in java](https://www.docs.opencv.org/java/2.4.9/)

I think `vpImage<char>`, `vpImage<vpRGBa>` and `vpArray2D<double>` are most frequently used. So its better to add keep only these as `VpImage`, `VpImageRGBa` and `VpArray` java classes and ignore the rest template variants.

fspindle commented:
I agree

AKS1996 commented:
Hi @fspindle @s-trinh 

I'm sure you must be able to build `visp_java` module.

Do have a look at visp's `java` module. In the `visp/modules/java/misc` directory you can find the modules with java support. Each folder in the directory contains a `filelist` document, which contains a list of header files/classes(like `include/visp3/vpColor.h` etc). Using it, we can specify java support for particular set of visp classes.

Currently I'm aiming at this simple [visp imgproc demo](http://visp-doc.inria.fr/doxygen/visp-daily/displayOpenCV_8cpp-example.html). I'll try to simplify it to make a working demo by tomorrow.

AKS1996 commented:
Sorry for responding late. I've been trying to figure what's the issue with `visp_java320.so` library - its not including functions for all modules. 

Nevertheless, I did a manual build for `core`, `imgproc`, `gui` and `io` modules. As already mentioned, I've created some [Eclipse demo projects](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/visp-java-demos). I guess you'll be [run the manual script](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/JNI-Usage-Example) and import those projects for development and analysis.

One more update. As you'll see in the `vpOpencvDisplay` demo, there's some qt initialization issue due to which its not working. As @s-trinh said, may be we should use Java swing for now.

I've tried reading and writing images using `io` and it works well. Try to build and run the demo's and let me know of any issues.

Thanks 
Akshay

AKS1996 commented:
Can anyone help out why tests output files are not being generated on Travis CI?

AKS1996 commented:
Thanks @fspindle for making tests work. That's a lot of changes. It might have taken a week for me. From now on, I'll try to keep the commits short/specific and watch out for duplicated code, if any.

fspindle commented:
@AKS1996 You are welcome.
Duplicate code was introduced by myself ;-)

I don't know if you have seen that I have started a new tutorial in `doc/tutorial/java/tutorial-install-java.doc` that should be completed.

AKS1996 commented:
Hi @fspindle 
I need a little bit help; CMake is turning out to be confusing. For simplicity I've reduced `visp_java` module to depend on `visp_core` only. 

To compile `libvisp_java.so` library, we include some handwritten cpp files(like VpMatrix.cpp, VpImageUChar.cpp, etc) and some generated cpp files(I've simplified the code to include only VpTime.cpp). But demangling shows that the library is not including generated cpp files. I've checked that cpp files are generated at build time before compiling the library. 

Specifically, `modules/java/generator` generates the JNI and Java code and `modules/java/jni` generates `libvisp_java.so` file. I've referred some links [like this](https://stackoverflow.com/questions/32512872/cmake-executable-with-auto-generated-sources) but no help.

fspindle commented:
@AKS1996 On my mac I'm unable to build `visp_java`:
````
[100%] Linking CXX shared module ../../../lib/libvisp_java320.dylib
Undefined symbols for architecture x86_64:
  "vpColVector::setIdentity(double const&)", referenced from:
      _Java_org_visp_core_VpColVector_setIdentity_10 in VpColVector.cpp.o
      _Java_org_visp_core_VpColVector_setIdentity_11 in VpColVector.cpp.o
  "vpRowVector::setIdentity(double const&)", referenced from:
      _Java_org_visp_core_VpRowVector_setIdentity_10 in VpRowVector.cpp.o
      _Java_org_visp_core_VpRowVector_setIdentity_11 in VpRowVector.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [lib/libvisp_java320.dylib] Error 1
````

Do you succeed to build `visp_java` ?

AKS1996 commented:
No. I just got a MacOS VM. Got the same error while checking steps for installation doc. Looking into the issue.

fspindle commented:
Wait, I'm on the issue removing the deprecated functions.

fspindle commented:
I'm unable to push changes on a new branch on your fork.
Could you add me as developper

AKS1996 commented:
Sure. I've sent you a collaborator invite link

AKS1996 commented:
Have a look at the doc. 
+ I installed Fedora 28 for checking. I guess `g++` is not pre-installed on it. I had to install `gcc-c++`
+ On Fedora, I didn't install JDK from oracle. `ant` comes with an open-jdk dependency. I tried and code built fine
+ On MacOS, `brew cask install java` gives the same JDK as that on Oracle's download page

AKS1996 commented:
Ah! forgot to add doc for windows :smile:

AKS1996 commented:
I'll change the python script to keep tabs of 2 spaces and remove empty lines

AKS1996 commented:
@fspindle I guess removing the `extern "C"` encapsulation is messing with the libraries. I tried running [previous demos](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/visp-java-demos) but they show unsatisfied linking errors now. 

As a test, try moving the `extern "C"` command [here](https://github.com/lagadic/visp/pull/323/commits/f3a5e0964d7b3ea38fcc614a96de5e10ef246fc8#diff-ca38113b052c1f6ba58c87f5923a37f7R11) to the end of file. Compile it and the resultant library works fine as earlier.

I don't know what's exactly the issue, but its causing problem by mangling the member function's name.

fspindle commented:
Sorry.
Then try to move `} // extern "C"`at the end of the files

AKS1996 commented:
Strange. I didn't know that opencv doesn't have image display capabilities in its java module(refer [this](https://stackoverflow.com/questions/16494916/equivalent-method-for-imshow-in-opencv-java-build) for more). Nevertheless, I' almost done adding a simple Java Swing tutorial for display `vpImage<uchar>`

AKS1996 commented:
Hi @fspindle. Have a look at visp_io module demo [eclipse project here](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/visp-java-demos/visp-io-module-demo)

AKS1996 commented:
Also get the `visp_gui` module demo [eclipse project here](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/visp-java-demos/visp-gui-module-demo). As @s-trinh recommended, I've used Java Swing for GUI - it's simple, platform independent and many developers will find it easy to use it.

Note that I haven't wrapped `visp_gui` for Java support. I think Android developers won't need `visp_gui` or `visp_io` modules at all (in fact the `getPixels` function in `VpImageUChar` and `VpImageRGBa` classes will suffice in Android).

I guess we should add these set of Java tutorials to the `doc/tutorial/java` folder.

fspindle commented:
I don't know what I'm doing wrong, but I'm unable to run [vpCore-Demo-Manual](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/visp-java-demos/vpCore-Demo-Manual)
```
Exception in thread "main" java.lang.Error: Unresolved compilation problems: 
	VpCameraParameters cannot be resolved to a type
	VpCameraParameters cannot be resolved to a type
	at Main.main(Main.java:39)
```

After a fresh ViSP build, what I have done in Eclipse:
- create a user library adding `.../visp-build/bin/visp-320.jar` external jar and set native library location to `.../visp-build/lib`
- open project `vpCore-Demo-Manual` from file system
- press button run

If I comment line 39 that brings the error I got an other one with VpMatrix
```
Exception in thread "main" java.lang.UnsatisfiedLinkError: org.visp.core.VpMatrix.n_vpMatrix(IID)J
	at org.visp.core.VpMatrix.n_vpMatrix(Native Method)
	at org.visp.core.VpMatrix.<init>(VpMatrix.java:26)
	at Main.main(Main.java:22)
```

Did I miss something ?

AKS1996 commented:
Strange. Can you see `libvisp_java320.so` in `.../visp-build/lib` folder?

fspindle commented:
Yes. I'm on OSX and I have `libvisp_java320.dylib` in `.../visp-build/lib`

s-trinh commented:
@AKS1996 This looks great!

I think you can add `frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);` in `VpDisplay` to really close the window otherwise the application looks still running in Eclipse.

Maybe `public String dump()` could be replaced by the overrided `public String toString()`? I think this should allow using something like:
```
VpImageRGBa colorImage = new VpImageRGBa(3, 5, new VpRGBa((char)255,(char)0,(char)0,(char)255));
System.out.println(colorImage);
```
and `public static void print(Object ... objs)` could be removed also?

AKS1996 commented:
@fspindle Are you able to run the demo now? I couldn't find anything except threads like [this](https://stackoverflow.com/questions/1124788/java-unresolved-compilation-problem). Anyway try to clean the project and re-run

fspindle commented:
Yes I succeed cleaning the project ;-) 
It was nice to see that all your 3 [demos](https://github.com/AKS1996/GSOC-Prep/tree/simple_graphics/visp-java-demos) were working on OSX and Ubuntu 16.04.

Current evaluation period will end July 9. Regarding your [proposal](https://github.com/AKS1996/GSOC-Prep/blob/master/Proposal/Proposal.pdf) we would be interested:
- to have a merge commit of PR #323 as soon as your work arround `visp_java` with existing bindings (`visp_core`, `visp_gui`, `visp_io`) becomes stable. I think that we are very close. More than 100 files were modified yet. Waiting more will lead to too much changes to follow your work
- `visp_gui` binding could be improved with display functionalities of simple drawings (lines, cross) for Java
- then adding new java bindings for `visp_imgproc`, `visp_detection` (`vpApriltag`) could be a new PR for each
- next step would be to make a new java demo that uses an image as input, and detect the Apriltags

Then for Android we are interested :
- to have a tutorial for beginners to start with ViSP and Android (we are beginners ;-) It could be based on the [project](https://github.com/AKS1996/GSOC-Prep/tree/master/VispAndroidDemoApp) you already have done
- then to have the same demo (apriltag) but using the java bindings 
- a PR with Android toolchain potential changes + a demo with live images that are displayed + filtered using `visp_imgproc` or using `visp_core - vpImageFilter` like in this [tutorial](http://visp-doc.inria.fr/doxygen/visp-daily/tutorial-image-ios.html) could be also nice 

I don't know if this is ok for you or if you have an other planned schedule ?
If you want, there is also the possibility to have a skype to discuss about that.
Let us know...

AKS1996 commented:
First of all, thanks. It would've taken me a lot of time to draft such a concrete plan. 
I'm all good with this plan(actually I was going to discuss closing of PR #323 with you in a few days). I understand that by making smaller PR's, the code will be easy to review later and changes easy to track.

Just want to make sure - are all the things(mentioned above) to be done before 9th July. Coz then I'll have to speed up a bit.

fspindle commented:
No, all the plan is not to do before July 9, it was just a reminder. Do just your best

AKS1996 commented:
We should close the PR now. I've executed the tutorials on Windows, with MinGW and it worked well.
Note that
- The appveyor is missing ant. So the java module is not building in appveyor tests
- On Windows, to use the IO module u need to have libpng and libjpeg. These are provided by GNU as 32 bit binaries only. So we need to use 32 bit JDK and Eclipse only
- In order to have 64 bit installation, I guess we need to compile libpng source code to generate binaries

AKS1996 commented:
Once the PR is merged, I'll begin with simple GUI features - displayLine and others
